pkg/services/datasourceproxy/datasourceproxy.go

```diff
@@ -0,0 +1,85 @@
+package datasourceproxy
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"regexp"
+
+	"github.com/grafana/grafana/pkg/api/datasource"
+	"github.com/grafana/grafana/pkg/api/pluginproxy"
+	"github.com/grafana/grafana/pkg/infra/metrics"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/registry"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+func init() {
+	registry.RegisterService(&DatasourceProxyService{})
+}
+
+type DatasourceProxyService struct {
+	DatasourceCache        datasources.CacheService      `inject:""`
+	PluginRequestValidator models.PluginRequestValidator `inject:""`
+	PluginManager          plugins.Manager               `inject:""`
+	Cfg                    *setting.Cfg                  `inject:""`
+}
+
+func (p *DatasourceProxyService) Init() error {
+	return nil
+}
+
+func (p *DatasourceProxyService) ProxyDataSourceRequest(c *models.ReqContext) {
+	p.ProxyDatasourceRequestWithID(c, c.ParamsInt64(":id"))
+}
+
+func (p *DatasourceProxyService) ProxyDatasourceRequestWithID(c *models.ReqContext, dsID int64) {
+	c.TimeRequest(metrics.MDataSourceProxyReqTimer)
+
+	ds, err := p.DatasourceCache.GetDatasource(dsID, c.SignedInUser, c.SkipCache)
+	if err != nil {
+		if errors.Is(err, models.ErrDataSourceAccessDenied) {
+			c.JsonApiErr(http.StatusForbidden, "Access denied to datasource", err)
+			return
+		}
+		c.JsonApiErr(http.StatusInternalServerError, "Unable to load datasource meta data", err)
+		return
+	}
+
+	err = p.PluginRequestValidator.Validate(ds.Url, c.Req.Request)
+	if err != nil {
+		c.JsonApiErr(http.StatusForbidden, "Access denied", err)
+		return
+	}
+
+	// find plugin
+	plugin := p.PluginManager.GetDataSource(ds.Type)
+	if plugin == nil {
+		c.JsonApiErr(http.StatusInternalServerError, "Unable to find datasource plugin", err)
+		return
+	}
+
+	proxyPath := getProxyPath(c)
+	proxy, err := pluginproxy.NewDataSourceProxy(ds, plugin, c, proxyPath, p.Cfg)
+	if err != nil {
+		if errors.Is(err, datasource.URLValidationError{}) {
+			c.JsonApiErr(http.StatusBadRequest, fmt.Sprintf("Invalid data source URL: %q", ds.Url), err)
+		} else {
+			c.JsonApiErr(http.StatusInternalServerError, "Failed creating data source proxy", err)
+		}
+		return
+	}
+	proxy.HandleRequest()
+}
+
+var proxyPathRegexp = regexp.MustCompile(`^\/api\/datasources\/proxy\/[\d]+\/?`)
```

Comment:
You should be aware of this change done in recent PR for figuring out the proxy path. Perhaps something you have to change eventually to support proxying from other HTTP API requests than `/api/datasource/proxy/{id}`. `proxyPath` is probably something that in the future should be passed from the HTTP API handler down to this service?